source/state_representation/src/DigitalIOState.cpp

```diff
@@ -0,0 +1,133 @@
+#include "state_representation/DigitalIOState.hpp"
+
+#include "state_representation/exceptions/IONotFoundException.hpp"
+
+using namespace state_representation::exceptions;
+
+namespace state_representation {
+
+DigitalIOState::DigitalIOState() : IOState<bool>() {
+  this->set_type(StateType::DIGITAL_IO_STATE);
+}
+
+DigitalIOState::DigitalIOState(const std::string& name, unsigned int nb_ios) : IOState<bool>(name, nb_ios) {
+  this->set_type(StateType::DIGITAL_IO_STATE);
+  this->data_ = Eigen::Vector<bool, Eigen::Dynamic>::Zero(nb_ios);
+}
+
+DigitalIOState::DigitalIOState(const std::string& name, const std::vector<std::string>& io_names) :
+  DigitalIOState(name, io_names.size()) {
+  this->set_names(io_names);
+}
+
+DigitalIOState::DigitalIOState(const DigitalIOState& state) : DigitalIOState(state.get_name(), state.get_names()) {
+  if (state) {
+    this->set_data(state.data());
+  }
+}
+
+DigitalIOState DigitalIOState::Zero(const std::string& name, unsigned int nb_ios) {
+  DigitalIOState zero = DigitalIOState(name, nb_ios);
+  // as opposed to the constructor specify this state to be filled
+  zero.set_empty(false);
+  return zero;
+}
+
+DigitalIOState DigitalIOState::Zero(const std::string& name, const std::vector<std::string>& io_names) {
+  DigitalIOState zero = DigitalIOState(name, io_names);
+  // as opposed to the constructor specify this state to be filled
+  zero.set_empty(false);
+  return zero;
+}
+
+DigitalIOState DigitalIOState::Random(const std::string& name, unsigned int nb_ios) {
+  DigitalIOState random = DigitalIOState(name, nb_ios);
+  // set all the state variables to random
+  random.set_data(Eigen::Vector<bool, -1>::Random(random.get_size()));
```

Comment:
I assume the `-1`  is the template value that specifies dynamic sizing. I would prefer you use the special "keyword" `Eigen::Dynamic` for readability and to conform with their intended API